core/chains/solana/soltxm/txm.go

```diff
@@ -25,82 +36,341 @@ var (
 	lggr = lggr.Named("Txm")
 	return &Txm{
-		starter: utils.StartStopOnce{},
-		tc:      tc,
-		lggr:    lggr,
-		queue:   make(chan *solanaGo.Transaction, MaxQueueLen), // queue can support 1000 pending txs
-		stop:    make(chan struct{}),
-		done:    make(chan struct{}),
-		cfg:     cfg,
+		starter:  utils.StartStopOnce{},
+		lggr:     lggr,
+		queue:    make(chan queueMsg, MaxQueueLen), // queue can support 1000 pending txs
+		simQueue: make(chan queueMsg, MaxQueueLen), // queue can support 1000 pending txs
+		stop:     make(chan struct{}),
+		cfg:      cfg,
+		txs:      NewTxProcesses(chainID),
+		client:   NewLazyLoad(tc),
 	}
 }
 
 // Start subscribes to queuing channel and processes them.
 func (txm *Txm) Start(context.Context) error {
 	return txm.starter.StartOnce("solanatxm", func() error {
+		txm.done.Add(3) // waitgroup: tx retry, confirmer, simulator
 		go txm.run()
 		return nil
 	})
 }
 
 func (txm *Txm) run() {
-	defer close(txm.done)
+	defer txm.done.Done()
 	ctx, cancel := utils.ContextFromChan(txm.stop)
 	defer cancel()
+
+	// start confirmer + simulator
+	go txm.confirm(ctx)
+	go txm.simulate(ctx)
+
 	for {
 		select {
-		case tx := <-txm.queue:
-			// TODO: this section could be better optimized for sending TXs quickly
-			// fetch client
-			client, err := txm.tc()
-			if err != nil {
-				txm.lggr.Errorw("failed to get client", "err", err)
-				continue
-			}
+		case msg := <-txm.queue:
 			// process tx
-			sig, err := client.SendTx(ctx, tx)
+			sig, err := txm.sendWithRetry(ctx, msg.tx, msg.timeout)
 			if err != nil {
-				txm.lggr.Criticalw("failed to send transaction", "err", err)
-				continue
+				txm.lggr.Errorw("failed to send transaction", "error", err)
+				txm.client.Clear() // clear client if tx fails immediately (potentially bad RPC)
+				continue           // skip remainining
 			}
-			txm.lggr.Debugw("successfully sent transaction", "signature", sig.String())
+
+			// send tx + signature to simulation queue
+			msg.signature = sig
+			select {
+			case txm.simQueue <- msg:
+			default:
+				txm.lggr.Warnw("failed to enqeue tx for simulation", "queueFull", len(txm.queue) == MaxQueueLen, "tx", msg)
+			}
+
+			txm.lggr.Debugw("transaction sent", "signature", sig.String())
 		case <-txm.stop:
 			return
 		}
 	}
 }
 
-// TODO: transaction confirmation
-// use ConfirmPollPeriod() in config
+func (txm *Txm) sendWithRetry(chanCtx context.Context, tx *solanaGo.Transaction, timeout time.Duration) (solanaGo.Signature, error) {
+	// fetch client
+	client, err := txm.client.Get()
+	if err != nil {
+		return solanaGo.Signature{}, errors.Wrap(err, "failed to get client in soltxm.sendWithRetry")
+	}
+
+	// create timeout context
+	ctx, cancel := context.WithTimeout(chanCtx, timeout)
+
+	// send initial tx (do not retry and exit early if fails)
+	sig, err := client.SendTx(ctx, tx)
+	if err != nil {
+		cancel()            // cancel context when exiting early
+		txm.txs.Failed(sig) // increment failed metric
```

Comment:
in this case, `txm.txs.Failed` wouldn't find the txhash (since it's technically the null/zero hash - just all `1`s) so it would return early before actually calling a cancelFunc or deleting the key